satellite/satellitedb/usercredits.go

```diff
@@ -0,0 +1,142 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package satellitedb
+
+import (
+	"context"
+	"time"
+
+	"github.com/skyrings/skyring-common/tools/uuid"
+	"github.com/zeebo/errs"
+
+	"storj.io/storj/satellite/console"
+	dbx "storj.io/storj/satellite/satellitedb/dbx"
+)
+
+type usercredits struct {
+	db *dbx.DB
+}
+
+// TotalReferredCount returns the total amount of referral a user has made based on user id
+func (c *usercredits) TotalReferredCount(ctx context.Context, id uuid.UUID) (int64, error) {
+	totalReferred, err := c.db.Count_UserCredit_By_ReferredBy(ctx, dbx.UserCredit_ReferredBy(id[:]))
+	if err != nil {
+		return totalReferred, errs.Wrap(err)
+	}
+
+	return totalReferred, nil
+}
+
+// GetAvailableCredits returns all records of user credit that are not expired or used
+func (c *usercredits) GetAvailableCredits(ctx context.Context, referrerID uuid.UUID, expirationEndDate time.Time) ([]console.UserCredit, error) {
+	availableCredits, err := c.db.All_UserCredit_By_UserId_And_ExpiresAt_Greater_And_CreditsUsedInCents_Less_CreditsEarnedInCents_OrderBy_Asc_ExpiresAt(ctx,
+		dbx.UserCredit_UserId(referrerID[:]),
+		dbx.UserCredit_ExpiresAt(expirationEndDate),
+	)
+	if err != nil {
+		return nil, errs.Wrap(err)
+	}
+
+	return fromDBX(availableCredits)
+}
+
+// Create insert a new record of user credit
+func (c *usercredits) Create(ctx context.Context, userCredit console.UserCredit) (*console.UserCredit, error) {
+	createdUser, err := c.db.Create_UserCredit(ctx,
+		dbx.UserCredit_UserId(userCredit.UserID[:]),
+		dbx.UserCredit_OfferId(userCredit.OfferID),
+		dbx.UserCredit_CreditsEarnedInCents(userCredit.CreditsEarnedInCents),
+		dbx.UserCredit_ExpiresAt(userCredit.ExpiresAt),
+		dbx.UserCredit_Create_Fields{
+			ReferredBy: dbx.UserCredit_ReferredBy(userCredit.ReferredBy[:]),
+		},
+	)
+	if err != nil {
+		return nil, errs.Wrap(err)
+	}
+
+	return convertDBX(createdUser)
+}
+
+// UpdateAvailableCredits updates user's available credits based on their spending and the time of their spending
+func (c *usercredits) UpdateAvailableCredits(ctx context.Context, appliedCredits int, id uuid.UUID, expirationEndDate time.Time) error {
+	tx, err := c.db.Open(ctx)
+	if err != nil {
+		return errs.Wrap(err)
+	}
+
+	availableCredits, err := tx.All_UserCredit_By_UserId_And_ExpiresAt_Greater_And_CreditsUsedInCents_Less_CreditsEarnedInCents_OrderBy_Asc_ExpiresAt(ctx,
+		dbx.UserCredit_UserId(id[:]),
+		dbx.UserCredit_ExpiresAt(expirationEndDate),
+	)
+	if err != nil {
+		return errs.Wrap(errs.Combine(err, tx.Rollback()))
+	}
+
+	for _, credit := range availableCredits {
+		if appliedCredits == 0 {
```

Comment:
`appliedCredits -= credit.CreditsEarnedInCents - credit.CreditsUsedInCents` could be less than 0?

Reply:
The statement on line 84 should make sure that never happen